### PR TITLE
*: fix system config for tenant

### LINKED ISSUE #75864: sql,config: tenant zone config lookup is broken
**Describe the problem**

In https://github.com/cockroachdb/cockroach/pull/74612 we made it so that the tenant has access to its own zone configs via the SystemConfig object provided by the `systemconfigwatcher`. Unfortunately this did not go far enough to actually fully address https://github.com/cockroachdb/cockroach/issues/70558 because of this special case logic: 

https://github.com/cockroachdb/cockroach/blob/4b8258c4b4a8f89fe8a4ea87fc8ee7cd5c916b6d/pkg/config/system.go#L371-L385

**Solution**

We need to either remove the above logic or make it conditional on where the config is coming from and more generally make sure that the SystemConfig methods work for the tenant. 

Epic: CRDB-10489

Jira issue: CRDB-12864

### PR DESCRIPTION
**Pull request by ajwerner**
Follow-on from https://github.com/cockroachdb/cockroach/pull/76279.

Fixes https://github.com/cockroachdb/cockroach/issues/75864.

This commit adds a mechanism to combine the system config data of the tenant
with the data provided over the GossipSubscription from the system tenant.
It then plumbs a version into the zone config methods. In the mixed version
state, the tenant uses the existing override from the system tenant. After
the span config infrastructure has been activated, the tenant uses the
overrides they've set. This affects, realistically, just the GC job, and,
to a lesser extent, the optimizer.

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
This is more or less ready for a look. There are some questions about the mixed version tenant behavior and about the `spanconfigcomparedccl` tests. @arulajmani let's discuss Monday.

**Review by arulajmani — COMMENTED** [review]
This seems pretty close. Should we unskip some of the locality optimized search logic tests that were blocked on #75864 as part of this patch as well? 

I still haven't looked at `spanconfigcomparedccl` diffs, but publishing comments for everything else. 

> Reviewed 31 of 33 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-:-MzStMC96metjQoE0gj8:b-ln0mt8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @ajwerner)

___
*[pkg/config/system.go, line 418 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSpGIbErLh0Q63Di6-:-MzSpGIbErLh0Q63Di60:bd5jmee) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/config/system.go#L418)):*
> ```Go
> 	// configuration.
> 	//
> 	// TODO(XXX): If the reconciliation protocol is not active, and this is a
> ```

We should probably remove the "XXX" and keep this comment around before we merge. Do you think it's worth being a bit verbose and adding some words about how the host changing `RANGE TENANTS` fits into this ugliness? 
___
*[pkg/config/system.go, line 428 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzQBgjd03XgXJSZiopd:-MzQBgjd03XgXJSZiope:bug5ngi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/config/system.go#L428)):*
> ```Go
> 	if !codec.ForSystemTenant() &&
> 		(id == 0 || !version.IsActive(
> 			clusterversion.EnsureSpanConfigReconciliation,
> ```

nit: In `GetConfReader`, we gate on `EnableSpanConfigStore` -- should we use that here as well? 
___
*[pkg/sql/zone_config.go, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSuafi5CyaienGCDVK:-MzSuafi5CyaienGCDVL:b-pka6t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/sql/zone_config.go#L201)):*
> ```Go
> 		return nil, nil, false, err
> 	}
> 	if err = completeZoneConfig(zone, keys.SystemSQLCodec, zoneID, getKey); err != nil {
> ```

Should this change to codec as well? 
___
*[pkg/sql/gcjob/refresh_statuses.go, line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSnrZO-JME4ZU-zibD:-MzSnrZO-JME4ZU-zibE:b867jne) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/sql/gcjob/refresh_statuses.go#L344)):*
> ```Go
> 	tenantTTLSeconds := execCfg.DefaultZoneConfig.GC.TTLSeconds
> 	v := execCfg.Settings.Version.ActiveVersionOrEmpty(ctx)
> 	zoneCfg, err := cfg.GetZoneConfigForObject(keys.MakeSQLCodec(roachpb.MakeTenantID(tenID)), v, 0)
> ```

Should we instead be passing in `execCfg.SystemSQLCodec` here instead? Then, we won't have to special case for this 0 value in `GetZoneConfigForObject`. 

Separately, should we instead be using `TenantsRangesID` instead of 0 here? Unless I'm misremembering, a draft version of this used `TenantsRangesID` here; has our thinking changed since?
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSdCIX3eQ6iYi0tIm7:-MzSdCIX3eQ6iYi0tIm8:b-ku5vha) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go#L40)):*
> ```Go
> // TestTenantUpgrade exercises the case where a system tenant is in a
> // non-finalized version state and creates a tenant. The test ensures
> // that newly created tenant begins in that same version.
> ```

nit: "that the"

File: pkg/ccl/spanconfigccl/spanconfigcomparedccl/testdata/basic
Comment (irfansharif) [inline pkg/ccl/spanconfigccl/spanconfigcomparedccl/testdata/basic:61]: Looking only at these tests, what this diff indicates is that according to the system config span, the new two tables have configs with `ignore_strict_gc=true rangefeed_enabled=true`. Is that expected? Shouldn't be, right? Since they're not system tables.

**Review by ajwerner — COMMENTED** [review]
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-:-N-Of04ACwCKnH9x8_ke:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @irfansharif)

___
*[pkg/config/system.go, line 418 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSpGIbErLh0Q63Di6-:-N-Of-9M1fxawqoyKRvj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/config/system.go#L418)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We should probably remove the "XXX" and keep this comment around before we merge. Do you think it's worth being a bit verbose and adding some words about how the host changing `RANGE TENANTS` fits into this ugliness? 
</blockquote></details>

Done.
___
*[pkg/config/system.go, line 428 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzQBgjd03XgXJSZiopd:-N-Oezb3-ywruvoJGyjI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/config/system.go#L428)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: In `GetConfReader`, we gate on `EnableSpanConfigStore` -- should we use that here as well? 
</blockquote></details>

Done.
___
*[pkg/sql/zone_config.go, line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSuafi5CyaienGCDVK:-N-Oeyfz9UxtILguzjsy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/sql/zone_config.go#L201)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this change to codec as well? 
</blockquote></details>

Done.
___
*[pkg/sql/gcjob/refresh_statuses.go, line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSnrZO-JME4ZU-zibD:-N-OeENv3PGXT0DH16YB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/sql/gcjob/refresh_statuses.go#L344)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we instead be passing in `execCfg.SystemSQLCodec` here instead? Then, we won't have to special case for this 0 value in `GetZoneConfigForObject`. 

Separately, should we instead be using `TenantsRangesID` instead of 0 here? Unless I'm misremembering, a draft version of this used `TenantsRangesID` here; has our thinking changed since?
</blockquote></details>

Done.
___
*[pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzSdCIX3eQ6iYi0tIm7:-N-OeK2x3aPlKiL0H4gk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/ccl/kvccl/kvtenantccl/tenant_upgrade_test.go#L40)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: "that the"
</blockquote></details>

Done.
___
*[pkg/ccl/spanconfigccl/spanconfigcomparedccl/testdata/basic, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-MzCIM23AL-CXFrOUPti-r1-61:-N-OetiD7LkllvB5GaYv:b-g7b8kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6efa71d2de43f7e52c0ecf22882b4659bd20269c/pkg/ccl/spanconfigccl/spanconfigcomparedccl/testdata/basic#L61)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Looking only at these tests, what this diff indicates is that according to the system config span, the new two tables have configs with `ignore_strict_gc=true rangefeed_enabled=true`. Is that expected? Shouldn't be, right? Since they're not system tables.
</blockquote></details>

I think I fixed it with Arul's help

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo comment 

> Reviewed 9 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-:-N-PGmik4aVBadtBmLx3:b-913mtg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @arulajmani, and @irfansharif)

___
*[pkg/config/system.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-N-PFM1G0AcabI9Ub79A:-N-PFM1G0AcabI9Ub79B:bfgfi3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff63efc516adc5a73b5d8471e13a5a76c6154f89/pkg/config/system.go#L403)):*
> ```Go
> 
> // GetZoneConfigForObject returns the combined zone config for the given object
> // identifier and SQL codec. Note that id=0 with a secondary tenant codec
> ```

Now that you've changed the GC code to use `RANGE TENANTS` id, can we get rid of this special casing?

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76504)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @arulajmani, and @irfansharif)

___
*[pkg/config/system.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76504#-N-PFM1G0AcabI9Ub79A:-N-PMGf57tUUCfvz1Aas:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff63efc516adc5a73b5d8471e13a5a76c6154f89/pkg/config/system.go#L403)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Now that you've changed the GC code to use `RANGE TENANTS` id, can we get rid of this special casing?
</blockquote></details>

Done.

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**
